### ISSUE TITLE
Memory leak in dmap library

### ISSUE THREAD

**Issue by asreimer** [vtsuperdarn/davitpy#191]
![v5qo0](https://cloud.githubusercontent.com/assets/4604819/11575369/36a9622a-99d6-11e5-9903-33db13c2b5ca.jpg)

Welp, there is a memory leak in the dmap library fo sho. The python wrappers are leaking memory like there's no tomorrow and it's awful when you try to process large sets of data (ie, more than a couple files at a time).

Every time a call is made to any of the functions in the dmapio.c file, RAM is used up an never returned.

For example, running the `rewind()` method uses up about 1 kilobyte of RAM each time it is run. Verify this yourself by running the following code:

```
from davitpy import pydarn
from datetime import datetime
import matplotlib

myPtr = pydarn.sdio.radDataOpen(datetime(2013,10,1),'cly',fileType='fitacf')
start_mem=matplotlib.cbook.report_memory()
myPtr.rewind()
end_mem=matplotlib.cbook.report_memory()
diff_mem = end_mem-start_mem

loop_num=100

start_mem=matplotlib.cbook.report_memory()
for i in range(0,loop_num):
    myPtr.rewind()

end_mem=matplotlib.cbook.report_memory()
print "Running rewind once required " + str(diff_mem) + " kilobytes of RAM"
print "Loop test started with " + str(start_mem) + " kilobytes of RAM"
print "Loop test ended with " + str(end_mem) + " kilobytes of RAM"
diff_mem = end_mem - start_mem
print "Loop test used " + str(diff_mem) + " kilobytes of RAM"
print "Loop test looped " + str(loop_num) +" times"
```

If you run through a bunch of files over and over again, it uses a ton of RAM and never returns it. Looping over the same file 20 times, I observed 75MB of RAM get chewed up. Here's code to verify it yourself:

```
from davitpy import pydarn
from datetime import datetime
import matplotlib

start_mem=matplotlib.cbook.report_memory()
myPtr = pydarn.sdio.radDataOpen(datetime(2013,10,1),'cly',fileType='fitacf')
myBeam = myPtr.readRec()
myBeam = myPtr.readRec()
end_mem=matplotlib.cbook.report_memory()
diff_mem = end_mem-start_mem

loop_num=20

start_mem=matplotlib.cbook.report_memory()
for i in range(0,loop_num):
    myPtr = pydarn.sdio.radDataOpen(datetime(2013,10,1),'cly',fileType='fitacf')
    myBeam = myPtr.readRec()
    while myBeam is not None:
        myBeam = myPtr.readRec()
end_mem=matplotlib.cbook.report_memory()
print "Running readRec twice required " + str(diff_mem) + " kilobytes of RAM"
print "Loop test started with " + str(start_mem) + " kilobytes of RAM"
print "Loop test ended with " + str(end_mem) + " kilobytes of RAM"
diff_mem = end_mem - start_mem
print "Loop test used " + str(diff_mem) + " kilobytes of RAM"
print "Loop test looped " + str(loop_num) +" times"
```

If I replace [this line](https://github.com/vtsuperdarn/davitpy/blob/develop/davitpy/pydarn/sdio/radDataTypes.py#L573) in the `radDataTypes.py` file with this:

```
offset=self.__ptr.tell()
```

which uses a built-in python method to "tell" the file offset, instead of using the `dmapio.c` `getDmapOffset` function, then the memory usage from the above test drops down to only 6MB for 20 loops.

This is kind of a big deal. It means that one cannot do any statistical studies of SuperDARN data using davitpy without some nasty work-arounds implemented.

The way to fix this is to fix the memory leak issues in the `dmapio.c` file or to replace all dmap C code with python (definitely doable!).

**Comment by aburrell** [comment]
This has been a pain in my bum for ages.  And yet I haven’t fixed it yet… ideally I’m happy to help fix it...

**Comment by ksterne** [comment]
Hey @asreimer,


Good to hear that someone else is starting to feel the pain as well.  I've been running code similar to what I wrote in #119 (and I feel like somewhere else, but I can't find it easily) and I can not get through a month of data for several radars, but it stops being able to read in files after some many radars (the loop goes through all of the radars in the network).  I agree that this is an issue if anyone wants to do some kind of statistical study this the code.  

Sorry I don't have much more to add on places to start digging or any suggestions for a solution.  Just know you're not alone in the bum hurt.

**Comment by asreimer** [comment]
I've heard that the people at UAF have a python version of the dmap library. I've emailed Jeff about it but haven't heard a reply yet. @ksterne do you have contact info for UAF's Engineer? If I get the chance I'm going to ask Bristow about it here at AGU.

**Comment by mikeruo** [comment]
There's a new 'Jef' at UAF, maybe Bill can help (cc'ed here).

Mike

**Comment by ajribeiro** [comment]
There had been a memory leak when I first wrote this, but I thought I had
fixed it.  It looks like it was refactored, but the issue is likely that
Py_CLEAR isn't being called every time it should be.  Should be fairly
straightforward to track down and fix

**Comment by asreimer** [comment]
Jef touched the dmap stuff after you wrote it. Maybe he introduced this problem. I'll have a look at the before and after. That Py_CLEAR advice gives me a starting point, thanks!

**Comment by asreimer** [comment]
Doesn't seem like there is any difference between @ajribeiro's version and what we currently have right now (except for the offsetSet and offsetTell functions that I previously mentioned). 

Not sure what is causing the memory leak in `read_dmap_rec`....

**Comment by asreimer** [comment]
Alright, so C is dead (long live C). Keith, our SuperDARN software engineer here in Saskatoon, built a purely python version of dmap that is basically as fast as the C wrapped version. I'll do some testing to see how comparable.

So ya, soon, no more RST dependence for davitpy! :-D

**Comment by asreimer** [comment]
Couldn't help it. Had to test the timing between the two. Keith's code is faster, even though it's written in pure python! Well it uses numpy which has C wrapped stuff...

Here are the results:
Current C wrapped dmap library (time to read an entire rawacf file into a record dictionaries):
2.2705s

@kkotyk's python based dmap library (time to read an entire rawacf file into a record dictionaries):
1.513s

There's only ~0.1s variance on multiple runs too. 

Here's the code I used to test our C wrapped library:


    import davitpy
    from datetime import datetime

    if __name__ == "__main__":
        stime = datetime.now()
        myPtr = davitpy.pydarn.sdio.radDataOpen(datetime(2015,8,31),'bks',fileType='rawacf',fileName='20150831.0000.03.bks.rawacf')
        records = list()
        rec = davitpy.pydarn.dmapio.readDmapRec(myPtr._radDataPtr__fd)
        records.append(rec)

        while rec is not None:
            rec = davitpy.pydarn.dmapio.readDmapRec(myPtr._radDataPtr__fd)
            records.append(rec)
        records.pop(-1)
        print datetime.now()-stime

**Comment by ksterne** [comment]
@asreimer, @kkotyk,

I'll gladly help test out this new version and see if it still gives me problems with my code that I'm running!

**Comment by ksterne** [comment]
Hey @asreimer, @kkotyk, any chance of getting the python dmap library into davitpy?  It'd be cool to see and I know there are many that would like to be able to use it.

**Comment by asreimer** [comment]
Keith is working on it right now actually. Once he has it all finished up then we can bug him about getting it merged in to davitpy.

It's pretty slick from what I've seen. I'm also looking forward to it being included.

**Comment by asreimer** [comment]
So @kkotyk is finished with what he has called `pydmap` now. I'm currently working on a branch to replace the RST C based dmap stuff in `dmapio` with `pydmap`. This could take some time.

**Comment by asreimer** [comment]
Fixed in #249, which replaces the RST C code with a python based dmap library written by @kkotyk.

**Comment by asreimer** [comment]
This was fixed in #249, but we never made a decision about it.
